keps/sig-auth/3299-kms-v2-improvements/README.md

```diff
@@ -0,0 +1,524 @@
+    end
+    kubeapiserver->>etcd: store encrypt response and encrypted DEK
+```
+
+#### Decrypt Request
+
+```mermaid
+sequenceDiagram
+    participant kubeapiserver
+    participant kmsplugin
+    participant externalkms
+    %% if local KEK in metadata, then using hierarchy
+    alt encrypted local KEK is in metadata
+      kubeapiserver->>kmsplugin: decrypt request <br/> {"ciphertext": "<encrypted DEK>", observed_key_id: "<current_key_id gotten as part of EncryptResponse>", <br/> "metadata": {"kms.kubernetes.io/local-kek": "<encrypted local KEK>"}}
+        alt encrypted local KEK in cache
+            kmsplugin->>kmsplugin: decrypt DEK with local KEK
+        else encrypted local KEK not in cache
+            kmsplugin->>externalkms: decrypt local KEK with remote KEK
+            externalkms->>kmsplugin: decrypted local KEK
+            kmsplugin->>kmsplugin: decrypt DEK with local KEK
+            kmsplugin->>kmsplugin: cache decrypted local KEK
+        end
+        kmsplugin->>kubeapiserver: return decrypt response <br/> {"plaintext": "<decrypted DEK>", current_key_id: "<remote KEK ID>", <br/> "metadata": {"kms.kubernetes.io/local-kek": "<encrypted local KEK>"}}
+    else encrypted local KEK is not in metadata
+        kubeapiserver->>kmsplugin: decrypt request <br/> {"ciphertext": "<encrypted DEK>", observed_key_id: "<current_key_id gotten as part of EncryptResponse>", <br/> "metadata": {}}
+        kmsplugin->>externalkms: decrypt DEK with remote KEK (same behavior as today)
+        externalkms->>kmsplugin: decrypted DEK
+        kmsplugin->>kubeapiserver: return decrypt response <br/> {"plaintext": "<decrypted DEK>", current_key_id: "<remote KEK ID>", <br/> "metadata": {}}
+    end
+```
+
+### Test Plan
+
+This section is incomplete and will be updated before the beta milestone.
+
+Performance:
+
+Unit tests covering:
+
+…
+
+Integration test covering:
+
+…
+
+Rotation:
+
+Unit tests covering:
+
+…
+
+Integration test covering:
+
+…
+
+Observability:
+
+Unit tests covering:
+
+…
+
+Integration test covering:
+
+…
+
+### Graduation Criteria
+
+Since the storage version API is still alpha, this KEP will simply aim to make it possible to have automated rotation when that API is enabled and has been updated to include the new fields.  The rotation feature will first be scoped to a single API server and will not be part of the graduation criteria for this KEP because the storage version API will take time to mature. However, testing of rotation will be part of the graduation criteria to confirm that the right information is being made available to allow for automated rotation when the storage version API graduates.
+
+#### Alpha
+
+- Feature implemented behind a feature flag
+- Initial unit and integration tests completed and enabled
+
+#### Beta
+
+- Gather feedback from providers using the feature
+- Any known bugs fixed
+
+#### GA
+
+- This is part of the KMS reference implementation
```

Comment:
What is the KMS reference implemenation? Can we leave Beta and GA graduation requirements as TBD so we don't need to settle on them now?